src/route.c

```diff
@@ -508,138 +508,108 @@ int route_flush(void)
-        return route_add(&cf->dst.in, cf->plen, flags,
-                         &cf->via.in, dev, &cf->src.in, cf->mtu, cf->metric);
+        return route_add(&dst->in, plen, flags,
+                     &gw->in, dev, &src->in, mtu, metric);
+
+    case DPVSAGENT_ROUTE_DEL:
+        /*fallthrough*/
     case SOCKOPT_SET_ROUTE_DEL:
-        return route_del(&cf->dst.in, cf->plen, flags,
-                         &cf->via.in, dev, &cf->src.in, cf->mtu, cf->metric);
+        return route_del(&dst->in, plen, flags,
+                     &gw->in, dev, &src->in, mtu, metric);
+
     case SOCKOPT_SET_ROUTE_SET:
         return EDPVS_NOTSUPP;
     case SOCKOPT_SET_ROUTE_FLUSH:
         return EDPVS_NOTSUPP;
     default:
         return EDPVS_NOTSUPP;
     }
+
+    return EDPVS_NOTSUPP;
 }
 
-static void route_fill_conf(int af, struct dp_vs_route_conf *cf,
-                           const struct route_entry *entry)
+static void route_fill_conf(int af, struct dp_vs_route_detail *detail, const struct route_entry *entry)
 {
-    memset(cf, 0, sizeof(*cf));
+    memset(detail, 0, sizeof(*detail));
+    detail->af = af;
+    detail->mtu = entry->mtu;
+    detail->metric = entry->metric;
+    detail->flags = entry->flag;
 
-    /*
-     * FIXME:
-     * some config fields are not implemented in route_entry.
-     */
-    cf->af      = af;
-    cf->dst.in  = entry->dest;
-    cf->plen    = entry->netmask;
-    cf->via.in  = entry->gw;
-    cf->src.in  = entry->src;
-    cf->mtu     = entry->mtu;
-    cf->metric  = entry->metric;
+    detail->dst.plen = entry->netmask;
 
-    if (entry->flag & RTF_LOCALIN){
-        cf->scope = ROUTE_CF_SCOPE_HOST;
-    } else if (entry->flag & RTF_KNI) {
-        cf->scope = ROUTE_CF_SCOPE_KNI;
-    } else if (entry->gw.s_addr == htonl(INADDR_ANY)) {
-        cf->scope = ROUTE_CF_SCOPE_LINK;
-        cf->flags |= ROUTE_CF_FLAG_ONLINK;
-    } else {
-        cf->scope = ROUTE_CF_SCOPE_GLOBAL;
-    }
+    detail->dst.addr.in = entry->dest;
+    detail->src.addr.in = entry->src;
+    detail->gateway.addr.in = entry->gw;
 
     if (entry->port)
-        snprintf(cf->ifname, sizeof(cf->ifname), "%s", entry->port->name);
-
-    return;
+        snprintf(detail->ifname, sizeof(detail->ifname), "%s", entry->port->name);
 }
 
 static int route_sockopt_get(sockoptid_t opt, const void *conf, size_t size,
                              void **out, size_t *outsize)
 {
-    const struct dp_vs_route_conf *cf;
     struct dp_vs_route_conf_array *array;
     size_t nroute, hash;
     struct route_entry *entry;
     struct netif_port *port = NULL;
     int off = 0;
-    bool outwall_table=false;
-
-    if (conf && size >= sizeof(*cf))
-        cf = conf;
-    else
-        cf = NULL;
-
-    if (cf && strlen(cf->ifname)) {
-        port = netif_port_get_by_name(cf->ifname);
+    char *ifname = ((struct dp_vs_route_detail*)conf)->ifname;
+    if (conf && strlen(ifname)) {
+        port = netif_port_get_by_name(ifname);
```

Comment:
`if (conf && strlen(ifname)) `:  the `conf` makes no sense in the if statements for it's already used in line:595.